### PR TITLE
[IBD patch 1] Fix block availability

### PR DESCRIPTION
**Pull request by ptschip**
This patch primarily fixes the issue of block availability during IBD.  We have to ensure that all connected or newly connected NETWORK nodes are assumed to have the blocks we nee

### CODE REVIEW

**Comment by ptschip** [comment]
test are failing in debug...getting some kind of timeout when the python scripts run.,,will investigate.

**Comment by ptschip** [comment]
lock order issue found and fixed...all test passing now.

File: src/main.cpp [hunk 1]
```diff
@@ -6398,6 +6399,16 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
             LOG(NET, "more getheaders (%d) to end to peer=%s (startheight:%d)\n", pindexLast->nHeight,
                 pfrom->GetLogName(), pfrom->nStartingHeight);
             pfrom->PushMessage(NetMsgType::GETHEADERS, chainActive.GetLocator(pindexLast), uint256());
+
+            // If we are here then we are beginning the process of IBD where we download all the headers.
+            // As a result we need to assume that every connected node is a full node and has all the
+            // blocks that we need.  Therefore, update block availability for every connected node. If we
```
Comment (gandrewstone) [inline src/main.cpp:6405]: Can we make this assumption?  What if they don't have all the blocks?  Can we instead ask for their best block hash?

Comment (ptschip) [reply src/main.cpp:6405]: @gandrewstone I think the assumption works for a few reasons:   

1) even if we asked for their best block they may not give it to us..or give us a fake one.
1a) how would we ask for their best block hash?  I don't know of any p2p message that
would we do that. 
2) this assumption is only in effect during IBD so we don't end up slowing down block transmission during normal operation.
3) most important: if we end up asking for a block from them and they don't have it, the request manager will just ask for it from another peer.  After 10 mins the block we first requested will get a block download timeout  and the peer will get disconnected.

If we don't do the above step then we'll end up downloading from only one peer unless/until we receive a header/inv for the next mined block which could be a minute or 2 hrs and by then IBD will be possibly very slow if it takes a while to get the next mined block header.

Comment (ptschip) [reply src/main.cpp:6405]: @gandrewstone perhaps what i could do is every time we download a batch of 2000 headers, then take the block hash from the last header and ask for that from each node....so with 500,000 blocks we would potentially ask for 250 headers from each of the connect peers which then would update their block availablility when/if the header is returned.  I'll try that out tomorrow and let you know how it goes.

**Comment by ptschip** [comment]
closing as all was in PR #970

**Pull request closed without merging**